Mathlib/Data/Matroid/Dual.lean

```diff
@@ -0,0 +1,264 @@
+
+    obtain ⟨e, ⟨(heB'' : e ∉ _), heE⟩, heI⟩ := exists_of_ssubset hssu
+    use e
+    simp_rw [mem_diff, insert_subset_iff, and_iff_left heI, and_iff_right heE, and_iff_right hIE]
+    refine' ⟨by_contra (fun heX ↦ heB'' (hB''₁ ⟨_, heI⟩)), ⟨B'', hB'', _⟩⟩
+    · rw [hX]; exact ⟨heE, heX⟩
+    rw [←union_singleton, disjoint_union_left, disjoint_singleton_left, and_iff_left heB'']
+    exact disjoint_of_subset_left hB''₂.2 disjoint_compl_left
+  indep_maximal := by
+    rintro X - I'⟨hI'E, B, hB, hI'B⟩ hI'X
+    obtain ⟨I, hI⟩ :=  M.exists_basis (M.E \ X)
+    obtain ⟨B', hB', hIB', hB'IB⟩ := hI.indep.exists_base_subset_union_base hB
+    refine' ⟨(X \ B') ∩ M.E,
+      ⟨_,subset_inter (subset_diff.mpr _) hI'E, (inter_subset_left _ _).trans (diff_subset _ _)⟩, _⟩
+    · simp only [inter_subset_right, true_and]
+      exact ⟨B', hB', disjoint_of_subset_left (inter_subset_left _ _) disjoint_sdiff_left⟩
+    · rw [and_iff_right hI'X]
+      refine' disjoint_of_subset_right hB'IB _
+      rw [disjoint_union_right, and_iff_left hI'B]
+      exact disjoint_of_subset hI'X hI.subset disjoint_sdiff_right
+    simp only [mem_setOf_eq, subset_inter_iff, and_imp, forall_exists_index]
+    intros J hJE B'' hB'' hdj _ hJX hssJ
+    rw [and_iff_left hJE]
+    rw [diff_eq, inter_right_comm, ←diff_eq, diff_subset_iff] at hssJ
+
+    have hI' : (B'' ∩ X) ∪ (B' \ X) ⊆ B'
+    · rw [union_subset_iff, and_iff_left (diff_subset _ _),
+      ←inter_eq_self_of_subset_left hB''.subset_ground, inter_right_comm, inter_assoc]
+
+      calc _ ⊆ _ := inter_subset_inter_right _ hssJ
+           _ ⊆ _ := by rw [inter_distrib_left, hdj.symm.inter_eq, union_empty]
+           _ ⊆ _ := inter_subset_right _ _
+
+    obtain ⟨B₁,hB₁,hI'B₁,hB₁I⟩ := (hB'.indep.subset hI').exists_base_subset_union_base hB''
+    rw [union_comm, ←union_assoc, union_eq_self_of_subset_right (inter_subset_left _ _)] at hB₁I
+
+    have : B₁ = B'
+    · refine' hB₁.eq_of_subset_indep hB'.indep (fun e he ↦ _)
+      refine' (hB₁I he).elim (fun heB'' ↦ _) (fun h ↦ h.1)
+      refine' (em (e ∈ X)).elim (fun heX ↦ hI' (Or.inl ⟨heB'', heX⟩)) (fun heX ↦ hIB' _)
+      refine' hI.mem_of_insert_indep ⟨hB₁.subset_ground he, heX⟩
+        (hB₁.indep.subset (insert_subset he _))
+      refine' (subset_union_of_subset_right (subset_diff.mpr ⟨hIB',_⟩) _).trans hI'B₁
+      refine' disjoint_of_subset_left hI.subset disjoint_sdiff_left
+
+    subst this
+
+    refine' subset_diff.mpr ⟨hJX, by_contra (fun hne ↦ _)⟩
+    obtain ⟨e, heJ, heB'⟩ := not_disjoint_iff.mp hne
+    obtain (heB'' | ⟨-,heX⟩ ) := hB₁I heB'
+    · exact hdj.ne_of_mem heJ heB'' rfl
+    exact heX (hJX heJ)
+  subset_ground := by tauto
+
+/-- The dual of a matroid; the bases are the complements (w.r.t `M.E`) of the bases of `M`. -/
+def dual (M : Matroid α) : Matroid α := M.dual_indepMatroid.matroid
+
+/-- A notation typeclass for matroid duality-/
+class Mdual (β : Type*) := (dual : β → β)
```

Comment:
Do you intend to have more than one instance of this class? If not, you could define the notation to use `Matroid.dual` directly and simplify things.